### PR TITLE
Implement Dynamodb with RDS fallback

### PR DESCRIPTION
**Pull request by kierenbeckett**
Changes to go with ONSDigital/eq-survey-runner#1635.  Create dynamodb tables and add new config variables

### CODE REVIEW

File: survey-runner-application/global_vars.tf [hunk 1]
```diff
@@ -161,6 +161,17 @@ variable "submitted_responses_table_name" {
...
+  description = "Table name of table used for storing questionnaire state"
+}
+
+variable "eq_session_table_name" {
+  description = "Table name of table used for storing user sessions"
+}
+variable "used_jti_claim_table_name" {
+  description = "Table name of table used for storing used JTI claims"
+}
+
```
Comment (jonnyshaw89) [inline survey-runner-application/global_vars.tf]: really petty, but inconsistent whitespace between `variable` blocks

File: survey-runner-application/survey_runner.tf [hunk 2]
```diff
@@ -232,6 +232,61 @@ resource "aws_elastic_beanstalk_environment" "survey_runner_prime" {
     name      = "EQ_SUBMITTED_RESPONSES_TABLE_NAME"
     value     = "${var.submitted_responses_table_name}"
   }
+  setting {
+    namespace = "aws:elasticbeanstalk:application:environment"
+    name      = "EQ_SUBMITTED_RESPONSES_DYNAMO_READ"
+    value     = "True"
```
Comment (jonnyshaw89) [inline survey-runner-application/survey_runner.tf]: This is defaulted to `True` so we shouldn't set to what is already default.
Also we shouldn't need this one at all

File: survey-runner.tf [hunk 3]
```diff
@@ -174,7 +221,37 @@ module "survey-runner-on-ecs" {
...
+              "dynamodb:DeleteItem"
+          ],
+          "Resource": "${module.survey-runner-dynamodb.eq_session_table_arn}"
+      },
+      {
+          "Sid": "",
+          "Effect": "Allow",
+          "Action": [
+              "dynamodb:PutItem",
+              "dynamodb:GetItem"
```
Comment (jonnyshaw89) [inline survey-runner.tf]: Technically this doesn't need `dynamodb:GetItem` but not sure if it will work without?

File: survey-runner.tf [hunk 4]
```diff
@@ -541,12 +618,18 @@ module "survey-runner-vpc" {
...
-  submitted_responses_read_capacity  = 1
-  submitted_responses_write_capacity = 1
+  submitted_responses_read_capacity  = 5
+  submitted_responses_write_capacity = 10
+  questionnaire_state_read_capacity  = 200
+  questionnaire_state_write_capacity = 150
+  eq_session_read_capacity  = 50
+  eq_session_write_capacity = 25
+  used_jti_claim_read_capacity  = 1
+  used_jti_claim_write_capacity = 25
```
Comment (jonnyshaw89) [inline survey-runner.tf]: These should go back to `1` before merge

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: survey-runner.tf (hunk 3 above)
Comment (kierenbeckett) [reply survey-runner.tf]: I'd be surprised if the `ConditionExpression` we use when putting works (or should work) without `GetItem`, going to leave it in to be safe.

Comment (jonnyshaw89) [reply survey-runner.tf]: i would wip it out ( ;) ) and give it a try?

Comment (kierenbeckett) [reply survey-runner.tf]: works without `GetItem` 😮

File: survey-runner-application/survey_runner.tf [hunk 5]
```diff
@@ -232,6 +232,51 @@ resource "aws_elastic_beanstalk_environment" "survey_runner_prime" {
...
   }
+  setting {
+    namespace = "aws:elasticbeanstalk:application:environment"
+    name      = "EQ_QUESTIONNAIRE_STATE_TABLE_NAME"
+    value     = "${var.questionnaire_state_table_name}"
+  }
+  setting {
+    namespace = "aws:elasticbeanstalk:application:environment"
+    name      = "EQ_QUESTIONNAIRE_STATE_DYNAMO_READ"
+    value     = "True"
```
Comment (jonnyshaw89) [inline survey-runner-application/survey_runner.tf]: If we hard code this then there is no way to switch it off for deployment?

File: survey-runner.tf [hunk 6]
```diff
@@ -137,6 +140,42 @@ module "survey-runner-on-ecs" {
         "name": "EQ_SUBMITTED_RESPONSES_TABLE_NAME",
         "value": "${module.survey-runner-dynamodb.submitted_responses_table_name}"
       },
+      {
+        "name": "EQ_QUESTIONNAIRE_STATE_TABLE_NAME",
+        "value": "${module.survey-runner-dynamodb.questionnaire_state_table_name}"
+      },
+      {
+        "name": "EQ_QUESTIONNAIRE_STATE_DYNAMO_READ",
+        "value": "True"
```
Comment (jonnyshaw89) [inline survey-runner.tf]: These should come from a variable to make testing easier

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: survey-runner.tf [hunk 7]
```diff
@@ -541,12 +615,26 @@ module "survey-runner-vpc" {
...
+  questionnaire_state_min_write_capacity = 1
+  questionnaire_state_max_write_capacity = 500
+  eq_session_min_read_capacity  = 1
+  eq_session_max_read_capacity  = 500
+  eq_session_min_write_capacity = 1
+  eq_session_max_write_capacity = 500
+  used_jti_claim_min_read_capacity  = 1
+  used_jti_claim_max_read_capacity  = 500
+  used_jti_claim_min_write_capacity = 1
+  used_jti_claim_max_write_capacity = 500
```
Comment (jonnyshaw89) [inline survey-runner.tf]: i would say 500 was a little high for dev default. maybe 25 max.
only really needs to be turned up for load testing which never normally happens on dev defaults

File: developer_defaults.tf [hunk 8]
```diff
@@ -187,7 +187,37 @@ variable "survey_runner_log_level" {
   default     = "INFO"
 }
 
-  # Survey Runner New Relic
+variable "survey_runner_questionnaire_state_dynamo_read" {
+  description = "Whether survey runner should read from DynmoDB for Questionnaire State objects"
+  default     = true
```
Comment (jonnyshaw89) [inline developer_defaults.tf]: Have you tried turning these off?
They don't work. they get passed to the app as `0` and `1`
We should specify them as `"True"` and `"False"`

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: developer_defaults.tf (hunk 8 above)
Comment (kierenbeckett) [reply developer_defaults.tf]: Good spot, thanks.

File: developer_defaults.tf [hunk 9]
```diff
@@ -159,7 +159,7 @@ variable "survey_runner_docker_registry" {
 
 variable "survey_runner_tag" {
   description = "The tag for the Survey Runner image to run"
-  default     = "latest"
+  default     = "eq-2158-implement-dynamodb-with-rds-fallback"
```
Comment (jonnyshaw89) [inline developer_defaults.tf]: We should change this back so we don't accidentally merge it

File: survey-runner.tf [hunk 10]
```diff
@@ -541,12 +615,27 @@ module "survey-runner-vpc" {
 }
 
 module "survey-runner-dynamodb" {
-  source                             = "github.com/ONSdigital/eq-terraform-dynamodb"
-  env                                = "${var.env}"
-  aws_access_key                     = "${var.aws_access_key}"
-  aws_secret_key                     = "${var.aws_secret_key}"
-  submitted_responses_read_capacity  = 1
-  submitted_responses_write_capacity = 1
+  source                                 = "github.com/ONSdigital/eq-terraform-dynamodb?ref=eq-2158-implement-dynamodb-with-rds-fallback"
```
Comment (jonnyshaw89) [inline survey-runner.tf]: One this module is in we can change this back

**Review by jonnyshaw89 — APPROVED** [review]

**Review by nealedj — APPROVED** [review]

**Comment by nealedj** [comment]
I think @MebinAbraham is also taking a look.

**Comment by MebinAbraham** [comment]
Seems to be working fine from a functional perspective. 👍

**Pull request merged**